parser/prism/Translator.cc

```diff
@@ -1086,7 +1086,51 @@ unique_ptr<parser::Node> Translator::translate(pm_node_t *node, bool preserveCon
 
             auto arg = translate(definedNode->value);
 
-            return make_unique<parser::Defined>(location.join(arg->loc), move(arg));
+            if (!directlyDesugar) {
+                return make_unique<parser::Defined>(location.join(arg->loc), move(arg));
+            }
+
+            ENFORCE(arg != nullptr);
+            ENFORCE(arg->hasDesugaredExpr());
+
+            // Desugar to `::Magic.defined_instance_var(ivar)` or `::Magic.defined_class_var(cvar)`
+            if (auto *ivar = parser::NodeWithExpr::cast_node<parser::IVar>(arg.get())) {
+                auto sym = MK::Symbol(arg->loc, ivar->name);
+                auto expr = MK::Send1(arg->loc, MK::Magic(arg->loc), core::Names::definedInstanceVar(),
+                                      location.copyWithZeroLength(), move(sym));
+                return make_node_with_expr<parser::Defined>(move(expr), location.join(arg->loc), move(arg));
+            }
+
+            if (auto *cvar = parser::NodeWithExpr::cast_node<parser::CVar>(arg.get())) {
+                auto sym = MK::Symbol(arg->loc, cvar->name);
+                auto expr = MK::Send1(arg->loc, MK::Magic(arg->loc), core::Names::definedClassVar(),
+                                      location.copyWithZeroLength(), move(sym));
+                return make_node_with_expr<parser::Defined>(move(expr), location.join(arg->loc), move(arg));
+            }
+
+            // Desugar to `::Magic.defined_p(a, b, c)` or `::Magic.defined_p()`
+            ast::Send::ARGS_store args;
+            if (parser::NodeWithExpr::isa_node<parser::Const>(arg.get())) {
```

Comment:
I think this conditional can go away: you're already asserting that there's a desugared expression above on line 1094, and the construction of the args vector is entirely dependent on the structure of that value, not the additional parser node.